lib/ui/channel_buffers.dart

```diff
@@ -6,184 +6,334 @@
+  /// a drain again), all in one stack frame (not allowing the drain
+  /// itself an opportunity to check if a listener is set).
+  bool _draining = false;
+
+  /// Adds a message to the channel.
+  ///
+  /// Returns true on overflow. Earlier messages are discarded,
+  /// in a first-in-first-out fashion. See [capacity].
+  bool push(_StoredMessage message) {
+    if (!_draining && _channelCallbackRecord != null) {
+      assert(_queue.isEmpty);
+      _channelCallbackRecord!.invoke(message.data, message.callback);
+      return false;
+    }
     if (_capacity <= 0) {
       return true;
-    } else {
-      final int overflowCount = _dropOverflowItems(_capacity - 1);
-      _queue.addLast(val);
-      return overflowCount > 0;
     }
+    final int overflowCount = _dropOverflowMessages(_capacity - 1);
+    _queue.addLast(message);
+    return overflowCount > 0;
   }
 
-  /// Returns null when empty.
-  T? pop() {
-    return _queue.isEmpty ? null : _queue.removeFirst();
-  }
+  /// Returns the first message in the channel and removes it.
+  ///
+  /// Throws when empty.
+  _StoredMessage pop() => _queue.removeFirst();
 
-  /// Removes items until then length reaches [lengthLimit] and returns
-  /// the number of items removed.
-  int _dropOverflowItems(int lengthLimit) {
+  /// Removes messages until [length] reaches `lengthLimit`, and returns
+  /// the number of messages removed.
+  int _dropOverflowMessages(int lengthLimit) {
     int result = 0;
     while (_queue.length > lengthLimit) {
-      final T item = _queue.removeFirst();
-      _dropItemCallback?.call(item);
+      final _StoredMessage message = _queue.removeFirst();
+      message.callback(null); // send empty reply to the plugin side
       result += 1;
     }
     return result;
   }
 
-  /// Returns the number of discarded items resulting from resize.
-  int resize(int newSize) {
-    _capacity = newSize;
-    return _dropOverflowItems(newSize);
+  _ChannelCallbackRecord? _channelCallbackRecord;
+
+  /// Sets the listener for this channel.
+  ///
+  /// When there is a listener, messages are sent immediately.
+  ///
+  /// If any messages were queued before the listener is added,
+  /// they are drained asynchronously after this method returns.
+  ///
+  /// Only one listener may be set at a time. Setting a
+  /// new listener clears the previous one.
+  ///
+  /// Callbacks are invoked in their own stack frame and
+  /// use the zone that was current when the callback was
+  /// registered.
+  void setListener(ChannelCallback callback) {
+    final bool needDrain = _channelCallbackRecord == null;
+    _channelCallbackRecord = _ChannelCallbackRecord(callback);
+    if (needDrain && !_draining)
+      _drain();
+  }
+
+  /// Clears the listener for this channel.
+  ///
+  /// When there is no listener, messages are queued, up to [capacity],
+  /// and then discarded in a first-in-first-out fashion.
+  void clearListener() {
```

Comment:
Clearing the listener while draining will lead to a null pointer exception.  I thought we said we would drop the messages.

Reply:
I will add a test for this case.

My intention was for removing a listener to just cause the messages to stop being drained, without losing them.

Reply:
test seems to confirm that clearListener() just pauses the draining.

Reply:
Ahh yep, thanks I missed the null check before its usage.